### ISSUE TITLE
VK_EXT_descriptor_indexing issues

### ISSUE THREAD

**Issue by MennoVink** [KhronosGroup/MoltenVK#1696]
I'm trying to use this extension but have ran into quite a few dead ends with mvk. My goal is to be able to use `layout( set = 0, binding = 1 ) uniform sampler2D albedoSamplers[];`.

I have this working without any validation issues on windows amd & nvidia with the following enabled:
- VK_EXT_DESCRIPTOR_INDEXING_EXTENSION_NAME
- vk::PhysicalDeviceFeatures::shaderSampledImageArrayDynamicIndexing
- vk::PhysicalDeviceFeatures::multiDrawIndirect
- vk::PhysicalDeviceDescriptorIndexingFeaturesEXT::runtimeDescriptorArray
- vk::PhysicalDeviceDescriptorIndexingFeaturesEXT::descriptorBindingVariableDescriptorCount

The issues i'm having on my mbp 2015 Intel Iris 6100 with mvk 1.3.216 on macos 12.4 are:
1. The first thing i note is that `vkDescriptorSetVariableDescriptorCountLayoutSupportEXT::maxVariableDescriptorCount` seems to be really low. mvk reports 30 as a maximum, but specifying 30 and `vk::DescriptorBindingFlagBitsEXT::eVariableDescriptorCount` still fails. Using 29 starts reporting the layout as supported, so it seems there's an off by one somewhere.
2. In order to try and raise the maximum of 29 samplers i tried to enable `MVKConfiguration::useMetalArgumentBuffers`. This is reporting errors to my debug utils callback. The error that mvk prints reads: `
2022-08-23 13:24:57.564374+0200 08_Bindless[2998:106789] Compiler failed to build request
[mvk-error] VK_ERROR_INITIALIZATION_FAILED: Render pipeline compile failed (Error code 2):
Unknown error found during compiling shader.
`
These are my shaders:
[obj_bindless.vert.txt](https://github.com/KhronosGroup/MoltenVK/files/9402088/obj_bindless.vert.txt)
[obj_bindless.frag.txt](https://github.com/KhronosGroup/MoltenVK/files/9402087/obj_bindless.frag.txt)
This probably affects #1689 as enabling that feature works for other example apps i made, but not for this one actually using bindless rendering.
3. Disabling `useMetalArgumentBuffers`, specifying max variable size of 29 but creating descriptor of 4 gives me this error when calling vkCreateGraphicsPipelines:
`
[mvk-error] VK_ERROR_INITIALIZATION_FAILED: Shader library compile failed (Error code 3):
program_source:30:167: error: cannot reserve 'sampler' resource locations at index 0
fragment main0_out main0(main0_in in [[stage_in]], const device buf& _21 [[buffer(1)]], array<texture2d<float>, 29> albedoSamplers [[texture(0)]], array<sampler, 29> albedoSamplersSmplr [[sampler(0)]])
                                                                                                                                                                      ^
.
`
4. Using variable array size of 6 (arbitrarily low), still with `vk::DescriptorBindingFlagBitsEXT::eVariableDescriptorCount` and the accompanying 4 in `vk::DescriptorSetVariableDescriptorCountAllocateInfoEXT` i am hitting program assertions inside vkCmdDrawIndexedIndirect. I am getting this metal validation error: `-[MTLDebugRenderCommandEncoder validateCommonDrawErrors:]:5252: failed assertion "Draw Errors Validation
Fragment Function(main0): missing sampler binding at index 4 for albedoSamplersSmplr[4].
Fragment Function(main0): missing sampler binding at index 5 for albedoSamplersSmplr[5].
"` The same issue happens when i use vkCmdDrawIndexed (no indirect buffer)
5. Having `useMetalArgumentBuffers` enabled, keeping maximum variable descriptor count arbitrarily low (ie 8 instead of 29) and trimming the amount of objects i'm rendering (to prevent gpu timeout on this slow gpu) does render at least something, but everything uses the albedo texture found at index 0. Enabling `vk::PhysicalDeviceDescriptorIndexingFeaturesEXT::shaderSampledImageArrayNonUniformIndexing` and using `nonuniformEXT( vInstanceIndex )` in the fragment shader has no effect.

So i gues my question is, is this extension expected to work? And are there any limitations that may not be listed yet? I couldn't find any mentions in `VK_KHR_portability_subset` or https://github.com/KhronosGroup/MoltenVK/blob/master/Docs/MoltenVK_Runtime_UserGuide.md#limitations
If this extension works for others, what may i be missing? Issue 4 looks like it fails to realize i'm using run-time variable array size. I'm providing `vk::DescriptorBindingFlagBitsEXT::eVariableDescriptorCount` during layout creation and `vk::DescriptorSetVariableDescriptorCountAllocateInfoEXT` during descriptor set creation. I think that's all i need to do right?

**Comment by billhollings** [comment]
Thanks for your diligent investigation.

`VK_EXT_descriptor_indexing` is currently in Beta, and is fragile on MoltenVK, due primarily to the mismatch between how Vulkan and Metal manage indexed resources. Metal 3 should improve this situation, once we roll it into descriptor indexing in MoltenVK.

However, that might not be the problem here.

Setting `useMetalArgumentBuffers` is the appropriate approach.

Unfortunately, the use of runtime arrays for `uniform sampler2D albedoSamplers[];`, is not supported because Metal dos not have the concept of combined image/samplers, and so MoltenVK needs to synthetically generate a Metal sampler for each image. This means that we end up with two runtime arrays...one for images and the other for samplers, and there is no way to pack two runtime arrays into a single resource struct.

So converting your fragment shader causes an error:

    MSL conversion error: Unsized array of images is not supported in MSL.

 I'm surprised you're not seeing that in your logs.

If you use a fixed-length array, something like `uniform sampler2D albedoSamplers[500];`, your fragment shader should compile correctly, with `useMetalArgumentBuffers` enabled.

BTW...if you set the `MVK_DEBUG=1` environment variable when you run your app, MoltenVK will log the shader conversions, and you'll be able to see how your GLSL shaders are converted to MSL.

**Comment by MennoVink** [comment]
I see, so do you want to get issues for bugs or will too much change with metal 3 anyway? ie i dont think a device should be allowed to open when runtimeDescriptorArray is requested but not supported.

As a side request, in the future, would it be possible to keep beta quality features outside of public release builds? Maybe keep them behind defines. That way people that want to play with them can build mvk themselves and test the functionality, but people that need to work with the library dont get tripped up in stuff that isn't reliable enough for production.

**Comment by billhollings** [comment]
> do you want to get issues for bugs

Definitely file bugs as you encounter them.

> i dont think a device should be allowed to open when `runtimeDescriptorArray` is requested but not supported.

Runtime arrays are supported for buffers, images, and samplers. But not for combined image/samplers, because of the behavioral difference with Metal, as I explained. This is the kind of difference that the Portability Initiative is striving to identify and accommodate.

> keep beta quality features outside of public release builds

Apologies for not documenting the status of these. We sometimes do formal betas when new features are being introduced.

**Comment by SRSaunders** [comment]
I recently submitted several PRs to Sascha Willem's Vulkan samples project (https://github.com/SaschaWillems/Vulkan), updating macOS and iOS compatibility/features across all of his examples for modern MoltenVK capabilities.  One of those was the descriptor indexing example, which now works well on macOS with argument buffers enabled.

FYI - I set the env variable at runtime by calling `setenv("MVK_CONFIG_USE_METAL_ARGUMENT_BUFFERS", "1", 1);` in the descriptor indexing example code.  This was simpler than using `vkSetMoltenVKConfigurationMVK()` due to an uncertain include path for the `vk_mvk_moltenvk.h` header file, which is installation dependent.

I also ran into a similar limits issue but with `maxPerStageDescriptorSamplers` being too small for the example (my device limit was 16 on an AMD RX580).  In this particular case I was able to solve it by creating the descriptor pool using the flag `VK_DESCRIPTOR_POOL_CREATE_UPDATE_AFTER_BIND_BIT`, activating a larger samplers limit (now 2048) named `maxPerStageDescriptorUpdateAfterBindSamplers` which was sufficient for the descriptor indexing example (32).

This descriptor indexing example won't work on iOS yet, but as @billhollings says above, perhaps that may be coming in the future by leveraging Metal 3 features in MoltenVK.

**Comment by billhollings** [comment]
In working on #1704, I experienced significant issues running CTS with Metal argument buffers enabled. Based on that, and above, we need to address the use of Metal argument buffers.

- Move to using Metal 3-style argument buffer building
- With Metal argument buffers enabled on M1, CTS currently crashes due to excessive GPU errors. This may be related to previous inability to run Metal argument shaders on iOS (ie- Apple Silicon).
- When working correctly, actually default `MVK_CONFIG_USE_METAL_ARGUMENT_BUFFERS` to `MVK_CONFIG_USE_METAL_ARGUMENT_BUFFERS_DESCRIPTOR_INDEXING`.

**Comment by SRSaunders** [comment]
Regarding your second point above, the symptoms sound similar to a problem I found with Sascha Wilems' _descriptorindexing_ example on M1.  The M1 GPU was reporting "GPU Soft Fault" errors when running this example on my M1 MacBook Air with Metal Argument Buffers enabled, but this did not happen on my Intel macOS desktop machine, nor on Windows or Linux.  At first I thought it was something to do with the M1 Vulkan implementation, but after much digging I found the following:

The _descriptorindexing_ example was referencing beyond the end of the Vertex buffer due a malformed index buffer.  The reason this showed up on the M1 and not my Intel box was due to differences in memory alignment between the two. Vulkan memory alignment block size on Intel is much larger than on M1, and when referencing beyond the nominal size of the Vertex buffer, it was actually still within the allocation block.  However, on M1 this was not the case and indexing beyond the end of the vertex buffer would actually run into non-allocated memory (or other allocations) - thus the GPU Soft Fault errors.  The solution in the end was generic - simply fix the construction of the Index buffer to not reference beyond the end of the vertex buffer.  See https://github.com/SRSaunders/Vulkan/commit/b60f404b0e613ffc7c5e4ee43d2d24ff3bb45607 for the indexing fix, which Sascha has now merged into main.

This may not be related to what you are observing with the CTS tests on M1, but I thought I would report it here in the off chance it could help.
